File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        
        ListNode * fast (head) ;
        ListNode* slow (head);
        while(n--){
            fast=fast->next;
        }
        if(fast==nullptr){
            return head->next;
        }
        ListNode* prev =nullptr;
        while(fast!=nullptr){
            prev = slow ;
            slow = slow-> next ; 
            fast = fast -> next ;
        }
        prev-> next = slow -> next ; 
        delete(slow);
        return head ;
        
    }
};


// head 
// tail delete 
// 

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
    let temp : ListNode | null =head;
    let k:number =0;
    let p:number =n;
    while(n>0&& temp!=null){
        temp=temp.next;
        k++;
        n--;
    }

    if(temp==null)return head.next
    let prev : ListNode | null =null;
    let root : ListNode | null = head;
    while(temp!=null){
        prev=root;
        root=root.next;
        temp=temp.next;

    }
    prev.next=root.next;
    return head;
};